File: apps/web/dashboard/services/clients.ts
import { api } from "@/services/api";

export type ClientLifecycleStatus = "lead" | "active" | "pending" | "inactive" | "blocked";
export type ClientCustomerType = "individual" | "business" | "agent" | "partner";
export type ClientSource = "manual" | "whatsapp" | "website" | "referral" | "import" | "api";

export type ClientRecord = {
  id: string;
  org_id: string;
  display_name: string | null;
  name: string | null;
  company_name: string | null;
  tax_id?: string | null;
  phone: string | null;
  whatsapp_phone: string | null;
  email: string | null;
  country: string | null;
  city: string | null;
  address?: string | null;
  customer_type: ClientCustomerType;
  lifecycle_status: ClientLifecycleStatus;
  source: ClientSource;
  preferred_language?: string | null;
  preferred_currency?: string | null;
  notes?: string | null;
  credit_enabled: boolean;
  credit_limit: number;
  current_balance: number;
  total_spent: number;
  payment_amount_due: number;
  payment_amount_paid: number;
  payment_currency: string | null;
  payment_status: "NOT_SET" | "UNPAID" | "PARTIAL" | "PAID";
  dossiers_count: number;
  shipments_count: number;
  packages_count?: number;
  last_activity_at: string | null;
  created_at: string;
  updated_at: string;
  row_version: number;
  deleted_at?: string | null;
};

export type ClientTimelineEvent = {
  id: string;
  type: "client" | "dossier" | "shipment" | "message" | "followup" | string;
  title: string;
  description: string;
  occurred_at: string;
  metadata: Record<string, unknown>;
};

export type ClientPackageSummary = {
  id: string;
  package_reference: string;
  tracking_id: string | null;
  status: string;
  weight_kg: number | null;
  destination_city: string | null;
  destination_country: string | null;
  received_at: string | null;
  dispatched_at: string | null;
  delivered_at: string | null;
  updated_at: string;
  departure_id: string | null;
  departure_code: string | null;
  departure_scheduled_at: string | null;
  departure_status: string | null;
};

export type ClientMessageSummary = {
  id: string;
  direction: "inbound" | "outbound";
  text_body: string | null;
  message_type: string | null;
  send_status: string | null;
  error_message: string | null;
  from_phone: string | null;
  to_phone: string | null;
  sender_name: string | null;
  is_group: boolean;
  media_mime_type: string | null;
  media_file_name: string | null;
  created_at: string;
};

export type ClientFinanceDocument = {
  id: string;
  document_type: "QUOTE" | "INVOICE" | "CREDIT_NOTE";
  document_number: string;
  status: string;
  currency: string;
  total: number;
  amount_paid: number;
  balance_due: number;
  issue_date: string | null;
  due_date: string | null;
  created_at: string;
};

export type ClientPaymentSummary = {
  id: string;
  receipt_number: string;
  amount: number;
  currency: string;
  method: string;
  reference: string | null;
  paid_at: string;
  status: string;
  document_id: string;
  document_number: string;
};

export type ClientWorkspace = {
  client: ClientRecord;
  packages: ClientPackageSummary[];
  messages: ClientMessageSummary[];
  documents: ClientFinanceDocument[];
  payments: ClientPaymentSummary[];
  summary: {
    packages: number;
    active_packages: number;
    messages: number;
    outstanding: number;
    paid: number;
  };
};

export type ClientDuplicate = Pick<
  ClientRecord,
  "id" | "display_name" | "name" | "company_name" | "phone" | "whatsapp_phone" | "email" | "country" | "city" | "customer_type" | "lifecycle_status" | "row_version"
> & {
  match_reason: "phone" | "email" | "name" | string;
  created_at: string;
};

export type ClientImportResult = {
  processed: number;
  created: number;
  skipped: number;
  errors: Array<{ row: number; error: string }>;
  clients: ClientRecord[];
};

export type ClientStats = {
  total: number;
  leads: number;
  active: number;
  pending: number;
  inactive: number;
  blocked: number;
  new_this_month: number;
};

export type ClientsResponse = {
  status: "ok";
  items: ClientRecord[];
  pagination: {
    page: number;
    page_size: number;
    total: number;
    total_pages: number;
  };
};

export type ClientPayload = {
  row_version?: number;
  name?: string;
  display_name?: string;
  company_name?: string;
  tax_id?: string;
  phone?: string;
  whatsapp_phone?: string;
  email?: string;
  country?: string;
  city?: string;
  address?: string;
  customer_type?: ClientCustomerType;
  lifecycle_status?: ClientLifecycleStatus;
  source?: ClientSource;
  preferred_language?: string;
  preferred_currency?: string;
  notes?: string;
  credit_enabled?: boolean;
  credit_limit?: number;
  payment_amount_due?: number;
  payment_amount_paid?: number;
  payment_currency?: string;
};

export async function listClients(params: {
  q?: string;
  status?: ClientLifecycleStatus | "";
  customer_type?: ClientCustomerType | "";
  source?: ClientSource | "";
  country?: string;
  city?: string;
  page?: number;
  page_size?: number;
  sort?: string;
} = {}) {
  return (await api.get<ClientsResponse>("/clients", { params })).data;
}

export async function getClient(id: string) {
  return (await api.get<{ status: "ok"; client: ClientRecord }>(`/clients/${id}`)).data.client;
}

export async function createClient(payload: ClientPayload) {
  return (await api.post<{ status: "ok"; client: ClientRecord }>("/clients", payload)).data.client;
}

export async function updateClient(id: string, payload: ClientPayload) {
  return (await api.patch<{ status: "ok"; client: ClientRecord }>(`/clients/${id}`, payload)).data.client;
}

export async function deleteClient(id: string, rowVersion: number) {
  return (await api.delete<{ status: "ok" }>(`/clients/${id}`, { params: { row_version: rowVersion } })).data;
}

export async function listArchivedClients(params: { q?: string; page?: number; page_size?: number } = {}) {
  return (await api.get<ClientsResponse>("/clients/archived", { params })).data;
}

export async function restoreClient(id: string, rowVersion: number) {
  return (await api.post<{ status: "ok"; client: ClientRecord }>(`/clients/${id}/restore`, null, { params: { row_version: rowVersion } })).data.client;
}

export async function mergeClients(payload: {
  source_client_id: string;
  target_client_id: string;
  source_version: number;
  target_version: number;
  idempotency_key: string;
}) {
  return (await api.post<{ status: "ok"; client: ClientRecord }>("/clients/merge", payload)).data.client;
}

export async function getClientStats() {
  return (await api.get<{ status: "ok"; stats: ClientStats }>("/clients/stats")).data.stats;
}

export async function getClientTimeline(id: string) {
  return (await api.get<{ status: "ok"; items: ClientTimelineEvent[] }>(`/clients/${id}/timeline`)).data.items;
}

export async function getClientWorkspace(id: string) {
  return (await api.get<{ status: "ok"; workspace: ClientWorkspace }>(`/clients/${id}/workspace`)).data.workspace;
}

export async function findClientDuplicates(params: {
  client_id?: string;
  phone?: string;
  email?: string;
  name?: string;
}) {
  return (await api.get<{ status: "ok"; items: ClientDuplicate[] }>("/clients/duplicates", { params })).data.items;
}

export async function importClients(file: File) {
  const formData = new FormData();
  formData.append("file", file);
  return (
    await api.post<{ status: "ok"; result: ClientImportResult }>("/clients/import", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    })
  ).data.result;
}

export async function exportClients(params: {
  q?: string;
  status?: ClientLifecycleStatus | "";
  customer_type?: ClientCustomerType | "";
  source?: ClientSource | "";
  country?: string;
  city?: string;
  sort?: string;
} = {}) {
  return (
    await api.get<Blob>("/clients/export", {
      params,
      responseType: "blob",
    })
  ).data;
}

File: apps/api/app/api/clients.py
import csv
import io

from fastapi import APIRouter, Depends, File, HTTPException, Query, Request, UploadFile, status
from pydantic import BaseModel, Field, model_validator
from starlette.responses import StreamingResponse

from app.audit.services.audit_service import audit_event
from app.clients.repository import (
    CLIENT_SOURCES,
    CLIENT_STATUSES,
    CLIENT_TYPES,
    client_timeline,
    client_workspace,
    client_stats,
    create_client,
    export_clients,
    find_client_duplicates,
    get_client,
    import_clients,
    list_clients,
    merge_clients,
    restore_client,
    soft_delete_client,
    update_client,
)
from app.core.tenant_context import get_current_tenant
from app.core.permissions import require_permission


router = APIRouter()
MAX_CLIENT_IMPORT_BYTES = 5 * 1024 * 1024
MAX_CLIENT_IMPORT_ROWS = 10_000
MAX_CLIENT_EXPORT_ROWS = 50_000


def csv_safe_value(value):
    """Prevent spreadsheet applications from evaluating exported cells as formulas."""
    if value is None:
        return ""
    rendered = str(value)
    if rendered.lstrip().startswith(("=", "+", "-", "@")):
        return f"'{rendered}"
    return rendered


class ClientPayload(BaseModel):
    name: str | None = Field(default=None, max_length=160)
    display_name: str | None = Field(default=None, max_length=180)
    company_name: str | None = Field(default=None, max_length=180)
    tax_id: str | None = Field(default=None, max_length=80)
    phone: str | None = Field(default=None, max_length=40)
    whatsapp_phone: str | None = Field(default=None, max_length=40)
    email: str | None = Field(default=None, max_length=180)
    country: str | None = Field(default=None, max_length=80)
    city: str | None = Field(default=None, max_length=80)
    address: str | None = Field(default=None, max_length=260)
    customer_type: str = "individual"
    lifecycle_status: str = "lead"
    source: str = "manual"
    preferred_language: str | None = Field(default="FR", max_length=10)
    preferred_currency: str | None = Field(default=None, max_length=12)
    notes: str | None = Field(default=None, max_length=2000)
    credit_enabled: bool = False
    credit_limit: float | None = 0
    current_balance: float | None = 0
    total_spent: float | None = 0
    payment_amount_due: float | None = 0
    payment_amount_paid: float | None = 0
    payment_currency: str | None = Field(default=None, min_length=3, max_length=3)

    @model_validator(mode="after")
    def validate_client(self):
        if not self.name and not self.company_name and not self.phone and not self.email:
            raise ValueError("name_company_phone_or_email_required")
        if self.customer_type not in CLIENT_TYPES:
            raise ValueError("invalid_customer_type")
        if self.lifecycle_status not in CLIENT_STATUSES:
            raise ValueError("invalid_lifecycle_status")
        if self.source not in CLIENT_SOURCES:
            raise ValueError("invalid_source")
        if (self.payment_amount_due or 0) < 0 or (self.payment_amount_paid or 0) < 0:
            raise ValueError("invalid_payment_amount")
        if self.payment_currency:
            self.payment_currency = self.payment_currency.upper()
        return self


class ClientPatchPayload(BaseModel):
    row_version: int = Field(ge=1)
    name: str | None = Field(default=None, max_length=160)
    display_name: str | None = Field(default=None, max_length=180)
    company_name: str | None = Field(default=None, max_length=180)
    tax_id: str | None = Field(default=None, max_length=80)
    phone: str | None = Field(default=None, max_length=40)
    whatsapp_phone: str | None = Field(default=None, max_length=40)
    email: str | None = Field(default=None, max_length=180)
    country: str | None = Field(default=None, max_length=80)
    city: str | None = Field(default=None, max_length=80)
    address: str | None = Field(default=None, max_length=260)
    customer_type: str | None = None
    lifecycle_status: str | None = None
    source: str | None = None
    preferred_language: str | None = Field(default=None, max_length=10)
    preferred_currency: str | None = Field(default=None, max_length=12)
    notes: str | None = Field(default=None, max_length=2000)
    credit_enabled: bool | None = None
    credit_limit: float | None = None
    current_balance: float | None = None
    total_spent: float | None = None
    payment_amount_due: float | None = None
    payment_amount_paid: float | None = None
    payment_currency: str | None = Field(default=None, min_length=3, max_length=3)

    @model_validator(mode="after")
    def validate_patch(self):
        if self.customer_type is not None and self.customer_type not in CLIENT_TYPES:
            raise ValueError("invalid_customer_type")
        if self.lifecycle_status is not None and self.lifecycle_status not in CLIENT_STATUSES:
            raise ValueError("invalid_lifecycle_status")
        if self.source is not None and self.source not in CLIENT_SOURCES:
            raise ValueError("invalid_source")
        if self.payment_amount_due is not None and self.payment_amount_due < 0:
            raise ValueError("invalid_payment_amount")
        if self.payment_amount_paid is not None and self.payment_amount_paid < 0:
            raise ValueError("invalid_payment_amount")
        if self.payment_currency:
            self.payment_currency = self.payment_currency.upper()
        return self


class ClientMergePayload(BaseModel):
    source_client_id: str = Field(min_length=1, max_length=64)
    target_client_id: str = Field(min_length=1, max_length=64)
    source_version: int = Field(ge=1)
    target_version: int = Field(ge=1)
    idempotency_key: str = Field(min_length=16, max_length=128)

    @model_validator(mode="after")
    def validate_merge(self):
        if self.source_client_id == self.target_client_id:
            raise ValueError("merge_same_client")
        return self


def _user_id(tenant: dict) -> str:
    return str(tenant.get("user_id") or "")


def _audit_client_bulk_operation(
    *, tenant: dict, request: Request, action: str, metadata: dict
) -> None:
    audit_event(
        org_id=tenant["org_id"],
        actor_id=_user_id(tenant),
        actor_name=tenant.get("actor_name") or _user_id(tenant),
        actor_role=tenant.get("actor_role"),
        entity_type="client_collection",
        entity_id=None,
        action=action,
        metadata=metadata,
        ip_address=request.client.host if request.client else None,
        user_agent=request.headers.get("user-agent"),
        severity="INFO",
    )


@router.get("/clients", dependencies=[Depends(require_permission("clients.read"))])
def clients_index(
    q: str | None = Query(default=None, max_length=120),
    status_filter: str | None = Query(default=None, alias="status"),
    customer_type: str | None = None,
    source: str | None = None,
    country: str | None = None,
    city: str | None = None,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    sort: str = "created_desc",
    tenant=Depends(get_current_tenant),
):
    if status_filter and status_filter not in CLIENT_STATUSES:
        raise HTTPException(status_code=422, detail="invalid_status")
    if customer_type and customer_type not in CLIENT_TYPES:
        raise HTTPException(status_code=422, detail="invalid_customer_type")
    if source and source not in CLIENT_SOURCES:
        raise HTTPException(status_code=422, detail="invalid_source")
    return {
        "status": "ok",
        **list_clients(
            tenant["org_id"],
            q=q,
            status=status_filter,
            customer_type=customer_type,
            source=source,
            country=country,
            city=city,
            page=page,
            page_size=page_size,
            sort=sort,
        ),
    }


@router.get("/clients/stats", dependencies=[Depends(require_permission("clients.read"))])
def clients_stats(tenant=Depends(get_current_tenant)):
    return {"status": "ok", "stats": client_stats(tenant["org_id"])}


@router.post("/clients/merge", dependencies=[Depends(require_permission("clients.merge"))])
def clients_merge(body: ClientMergePayload, tenant=Depends(get_current_tenant)):
    try:
        client = merge_clients(
            tenant["org_id"], body.source_client_id, body.target_client_id,
            _user_id(tenant), source_version=body.source_version,
            target_version=body.target_version, idempotency_key=body.idempotency_key,
        )
    except ValueError as exc:
        detail = str(exc)
        if detail in {"stale_client_version", "merge_relationship_conflict"}:
            raise HTTPException(status_code=409, detail=detail) from exc
        if detail in {"merge_client_not_found", "merge_target_not_found"}:
            raise HTTPException(status_code=404, detail=detail) from exc
        raise
    return {"status": "ok", "client": client}


@router.get(
    "/clients/archived",
    dependencies=[Depends(require_permission("clients.archive"))],
)
def clients_archived(
    q: str | None = Query(default=None, max_length=120),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    tenant=Depends(get_current_tenant),
):
    return {
        "status": "ok",
        **list_clients(
            tenant["org_id"], q=q, page=page, page_size=page_size,
            sort="activity_desc", archived=True,
        ),
    }


@router.post(
    "/clients",
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_permission("clients.create"))],
)
def clients_create(body: ClientPayload, tenant=Depends(get_current_tenant)):
    try:
        client = create_client(tenant["org_id"], _user_id(tenant), body.model_dump())
    except ValueError as exc:
        if str(exc) == "duplicate_client":
            raise HTTPException(status_code=409, detail="duplicate_client") from exc
        if str(exc) in {"invalid_phone", "invalid_email"}:
            raise HTTPException(status_code=422, detail=str(exc)) from exc
        raise
    return {"status": "ok", "client": client}


@router.get("/clients/export", dependencies=[Depends(require_permission("clients.export"))])
def clients_export(
    request: Request,
    q: str | None = Query(default=None, max_length=120),
    status_filter: str | None = Query(default=None, alias="status"),
    customer_type: str | None = None,
    source: str | None = None,
    country: str | None = None,
    city: str | None = None,
    sort: str = "created_desc",
    tenant=Depends(get_current_tenant),
):
    rows = export_clients(
        tenant["org_id"],
        q=q,
        status=status_filter,
        customer_type=customer_type,
        source=source,
        country=country,
        city=city,
        sort=sort,
        limit=MAX_CLIENT_EXPORT_ROWS + 1,
    )
    if len(rows) > MAX_CLIENT_EXPORT_ROWS:
        raise HTTPException(status_code=413, detail="client_export_too_large")
    _audit_client_bulk_operation(
        tenant=tenant,
        request=request,
        action="clients.exported",
        metadata={
            "row_count": len(rows),
            "filters": {
                "has_search": bool(q),
                "status": status_filter,
                "customer_type": customer_type,
                "source": source,
                "country": country,
                "city": city,
                "sort": sort,
            },
        },
    )
    output = io.StringIO(newline="")
    fieldnames = [
        "display_name",
        "name",
        "company_name",
        "phone",
        "whatsapp_phone",
        "email",
        "country",
        "city",
        "customer_type",
        "lifecycle_status",
        "source",
        "preferred_language",
        "preferred_currency",
        "credit_enabled",
        "credit_limit",
        "current_balance",
        "total_spent",
        "payment_amount_due",
        "payment_amount_paid",
        "payment_currency",
        "notes",
    ]
    writer = csv.DictWriter(output, fieldnames=fieldnames)
    writer.writeheader()
    for row in rows:
        writer.writerow({key: csv_safe_value(row.get(key)) for key in fieldnames})
    output.seek(0)
    return StreamingResponse(
        iter(["\ufeff" + output.getvalue()]),
        media_type="text/csv; charset=utf-8",
        headers={"Content-Disposition": 'attachment; filename="slaivio-clients.csv"'},
    )


@router.post("/clients/import", dependencies=[Depends(require_permission("clients.import"))])
async def clients_import(
    request: Request,
    file: UploadFile = File(...),
    tenant=Depends(get_current_tenant),
):
    if not file.filename or not file.filename.lower().endswith(".csv"):
        raise HTTPException(status_code=422, detail="csv_required")
    content = await file.read(MAX_CLIENT_IMPORT_BYTES + 1)
    if len(content) > MAX_CLIENT_IMPORT_BYTES:
        raise HTTPException(status_code=413, detail="client_import_too_large")
    try:
        decoded = content.decode("utf-8-sig")
    except UnicodeDecodeError as exc:
        raise HTTPException(status_code=422, detail="invalid_csv_encoding") from exc
    reader = csv.DictReader(io.StringIO(decoded, newline=""))
    if not reader.fieldnames:
        raise HTTPException(status_code=422, detail="empty_csv")
    normalized_headers = [str(header or "").strip().lower() for header in reader.fieldnames]
    if any(not header for header in normalized_headers) or len(set(normalized_headers)) != len(normalized_headers):
        raise HTTPException(status_code=422, detail="invalid_csv_headers")
    reader.fieldnames = normalized_headers
    rows = []
    for csv_row_number, row in enumerate(reader, start=2):
        if len(rows) >= MAX_CLIENT_IMPORT_ROWS:
            raise HTTPException(status_code=413, detail="client_import_too_many_rows")
        if None in row:
            raise HTTPException(status_code=422, detail="invalid_csv_row_shape")
        normalized_row = {str(key).strip().lower(): (value or "").strip() for key, value in row.items()}
        if not any(normalized_row.values()):
            continue
        normalized_row["_csv_row"] = csv_row_number
        rows.append(normalized_row)
    if not rows:
        raise HTTPException(status_code=422, detail="empty_csv")
    result = import_clients(tenant["org_id"], _user_id(tenant), rows)
    _audit_client_bulk_operation(
        tenant=tenant,
        request=request,
        action="clients.imported",
        metadata={
            "processed": result["processed"],
            "created": result["created"],
            "skipped": result["skipped"],
            "error_count": len(result["errors"]),
        },
    )
    return {"status": "ok", "result": result}


@router.get("/clients/duplicates", dependencies=[Depends(require_permission("clients.read"))])
def clients_duplicates(
    client_id: str | None = None,
    phone: str | None = None,
    email: str | None = None,
    name: str | None = None,
    tenant=Depends(get_current_tenant),
):
    return {
        "status": "ok",
        "items": find_client_duplicates(
            tenant["org_id"],
            client_id=client_id,
            phone=phone,
            email=email,
            name=name,
        ),
    }


@router.get("/clients/{client_id}", dependencies=[Depends(require_permission("clients.read"))])
def clients_show(client_id: str, tenant=Depends(get_current_tenant)):
    client = get_client(tenant["org_id"], client_id)
    if not client:
        raise HTTPException(status_code=404, detail="client_not_found")
    return {"status": "ok", "client": client}


@router.get(
    "/clients/{client_id}/timeline",
    dependencies=[Depends(require_permission("clients.read"))],
)
def clients_timeline(client_id: str, tenant=Depends(get_current_tenant)):
    client = get_client(tenant["org_id"], client_id)
    if not client:
        raise HTTPException(status_code=404, detail="client_not_found")
    return {"status": "ok", "items": client_timeline(tenant["org_id"], client_id)}


@router.get(
    "/clients/{client_id}/workspace",
    dependencies=[Depends(require_permission("clients.read"))],
)
def clients_workspace(client_id: str, tenant=Depends(get_current_tenant)):
    workspace = client_workspace(tenant["org_id"], client_id)
    if not workspace:
        raise HTTPException(status_code=404, detail="client_not_found")
    return {"status": "ok", "workspace": workspace}


@router.patch(
    "/clients/{client_id}",
    dependencies=[Depends(require_permission("clients.update"))],
)
def clients_update(client_id: str, body: ClientPatchPayload, tenant=Depends(get_current_tenant)):
    payload = body.model_dump(exclude_unset=True)
    try:
        client = update_client(tenant["org_id"], client_id, _user_id(tenant), payload)
    except ValueError as exc:
        if str(exc) == "duplicate_client":
            raise HTTPException(status_code=409, detail="duplicate_client") from exc
        if str(exc) == "stale_client_version":
            raise HTTPException(status_code=409, detail="stale_client_version") from exc
        if str(exc) in {"invalid_phone", "invalid_email"}:
            raise HTTPException(status_code=422, detail=str(exc)) from exc
        raise
    if not client:
        raise HTTPException(status_code=404, detail="client_not_found")
    return {"status": "ok", "client": client}


@router.delete(
    "/clients/{client_id}",
    dependencies=[Depends(require_permission("clients.archive"))],
)
def clients_delete(
    client_id: str,
    row_version: int = Query(ge=1),
    tenant=Depends(get_current_tenant),
):
    try:
        deleted = soft_delete_client(
            tenant["org_id"], client_id, _user_id(tenant), expected_version=row_version
        )
    except ValueError as exc:
        if str(exc) == "stale_client_version":
            raise HTTPException(status_code=409, detail="stale_client_version") from exc
        raise
    if not deleted:
        raise HTTPException(status_code=404, detail="client_not_found")
    return {"status": "ok"}


@router.post(
    "/clients/{client_id}/restore",
    dependencies=[Depends(require_permission("clients.archive"))],
)
def clients_restore(
    client_id: str,
    row_version: int = Query(ge=1),
    tenant=Depends(get_current_tenant),
):
    try:
        client = restore_client(
            tenant["org_id"], client_id, _user_id(tenant), expected_version=row_version
        )
    except ValueError as exc:
        if str(exc) == "restore_identity_conflict":
            raise HTTPException(status_code=409, detail="restore_identity_conflict") from exc
        if str(exc) == "stale_client_version":
            raise HTTPException(status_code=409, detail="stale_client_version") from exc
        raise
    if not client:
        raise HTTPException(status_code=404, detail="archived_client_not_found")
    return {"status": "ok", "client": client}
